example/src/main/java/com/klarna/example/MainActivity.java

```diff
@@ -93,14 +127,82 @@ public void openKlarnaPreferences(View view) {
         Intent intent = new Intent(this, PreferencesActivity.class);
         intent.putExtra(PreferencesActivity.EXTRA_USER_TOKEN, getUserToken());
         startActivityForResult(intent, PREFERENCES_REQUEST_CODE);
-
     }
 
-    public void onBuy(View view) {
+    public void onBuyPressed(View view) {
+        if (!this.hasUserToken()) {
+            openKlarnaRegistration();
+        }
+        
+        class purchaseItemRunnable implements Runnable {
+            String reference;
+            OriginProof originProof;
+
+            purchaseItemRunnable(String reference, OriginProof originProof) {
+                this.reference = reference;
+                this.originProof = originProof;
+            }
+
+            @Override
+            public void run() {
+                try {
+                    performPurchaseOfItem(reference, originProof);
+                } catch (final Exception e) {
+                    Handler mainHandler = new Handler(getApplicationContext().getMainLooper());
+
+                    mainHandler.post(new Runnable() {
+                        @Override
+                        public void run() {
+                            showAlert("Error: " + e.toString());
+                            Log.e(getClass().getName(), "Error on performPurchaseOfItem", e);
```

Comment:
This thing is still an "app". It shouldn't spew internal method names at unsuspecting users.